src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,447 @@
+      )
+      Future(scheduler ? ScheduleTrigger(PlanEnergyDispatchTrigger(0), self))
+        .map(_ => CompletionNotice(triggerId, Vector()))
+        .pipeTo(sender())
+
+    case TriggerWithId(PlanEnergyDispatchTrigger(timeBin), triggerId) =>
+      log.debug(s"Planning energy dispatch for vehicles currently connected to a charging point, at t=$timeBin")
+      import scala.concurrent.ExecutionContext.Implicits.global
+      import scala.concurrent.duration._
+      import scala.concurrent.{Await, Future, TimeoutException}
+      val estimatedLoad = requiredPowerInKWOverNextPlanningHorizon(timeBin)
+      log.debug("Total Load estimated is {} at tick {}", estimatedLoad.values.sum, timeBin)
+
+      // obtaining physical bounds
+      val physicalBounds = obtainPowerPhysicalBounds(timeBin, Some(estimatedLoad))
+
+      val futures = Future
+        .sequence(chargingNetworkMap.flatMap {
+          case (_, chargingNetwork) =>
+            chargingNetwork.connectedVehicles.map {
+              case (_, chargingVehicle @ ChargingVehicle(vehicle, stall, _, _, _, _)) =>
+                Future {
+                  // Refuel
+                  handleRefueling(chargingVehicle)
+                  // Calculate the energy to charge each vehicle connected to the a charging station
+                  val (chargingDuration, energyToCharge) =
+                    dispatchEnergy(cnmConfig.timeStepInSeconds, chargingVehicle, physicalBounds)
+                  // update charging vehicle with dispatched energy and schedule ChargingTimeOutScheduleTrigger
+                  chargingVehicle.processChargingCycle(timeBin, energyToCharge, chargingDuration).flatMap {
+                    case cycle if chargingIsCompleteUsing(cycle) =>
+                      handleEndCharging(timeBin, chargingVehicle)
+                      None
+                    case cycle if chargingNotCompleteUsing(cycle) =>
+                      log.debug(
+                        "Ending refuel cycle of vehicle {}. Stall: {}. Provided energy: {} J. Remaining: {} J",
+                        vehicle.id,
+                        stall,
+                        cycle.energy,
+                        energyToCharge
+                      )
+                      None
+                    case cycle =>
+                      // charging is going to end during this current session
+                      Some(
+                        ScheduleTrigger(
+                          ChargingTimeOutTrigger(timeBin + cycle.duration, vehicle.id, stall.managerId),
+                          self
+                        )
+                      )
+                  }
+                }
+            }
+        })
+        .map(_.toVector)
+        .recover {
+          case e =>
+            log.debug(s"No energy was dispatched at time $timeBin because: $e")
+            Vector.empty[Option[ScheduleTrigger]]
+        }
+
+      val triggers = (try {
+        Await.result(futures, atMost = 1.minutes)
```

Comment:
nitpick: instead of blocking you can use `pipeTo` to send the result back to yourself
https://doc.akka.io/docs/akka/2.5/futures.html#use-the-pipe-pattern

Reply:
this is much better. I replaced it with:
`        .map { triggers =>
          val nextStepPlanning = if (!isEndOfSimulation(timeBin)) {
            Vector(ScheduleTrigger(PlanEnergyDispatchTrigger(nextTimeBin(timeBin)), self))
          } else {
            completeChargingAndTheDisconnectionOfAllConnectedVehiclesAtEndOfSimulation(timeBin, physicalBounds)
          }
          CompletionNotice(triggerId, triggers.flatten.toVector ++ nextStepPlanning)
        }
        .pipeTo(sender())`